### DISCUSSION TITLE
Internal "bank" support for live performances with guitar rig emulation software (like AmpliTube)

### DISCUSSION THREAD

**Discussion by n3wkular** [shanteacontrols/OpenDeck#157] [Ideas]
This is similar to #155, but I am NOT talking about MIDI banks on the device we are controlling (like a synth); rather, I am talking about a logical "bank" internal to the controller. Let me try to explain ...

Consider this scenario (using AmpliTube as an example):

- AmpliTube allows the user to define 128 different guitar rig presets (NOT OpenDeck presets!)
- It also allows the user to program a MIDI controller to send PC commands to load presets. For example, sending PC#0 loads preset 001, sending PC#1 loads preset 002, and so on.
- I've built a controller with 4 momentary footswitches and everything works great when selecting between AmpliTube presets 001-004 (or whatever I've programmed to the four footswitches).
- I can also easily program 2 of the footswitches to increment and decrement through all 128 AmpliTube presets with no troubles.

However, this doesn't work well in a live setting:

- Going from preset 10 to 37 is a nightmare, for example, unless you've programmed them to one of the four buttons on the controller. 
- But that quickly becomes a problem because a guitarist uses far more than just 4 presets in a performance (20+ is more typical).
- And, no, I am not going to build a controller with 128 footswitches :grin: 

A pattern I've seen on guitar amp software controllers:

- A small-ish number of footswitches (4-10, usually) to send PC commands to select presets in the software.
- 2 footswitches that increment and decrement through logical "banks" on _the controller side_, _NOT_ in the device being controlled. (So, no MIDI message is being sent from these footswitches).
- **A logical "bank" in the controller is really just an index offset. For example, if our controller has 4 footswitches for PC messages, and we are on logical bank 0 in the controller, then pressing footswitch 1 would send PC#0. However, if we're on logical bank 1 in the controller, then pressing footswitch 1 would send PC#4. On logical bank 2, we'd get PC#8, and so on.**

This pattern works fairly well as long as the guitarist has arranged their presets so they rarely need to change logical banks during a song. Typically, each logical bank is used for a single song (or multiple songs that use the same presets) and the performer changes presets for the intro, chorus, solo, etc.

Now, if you've already thought of this and I just haven't figured out how to make OpenDeck do it, I'd be a happy man. But if OpenDeck can't already do this, what are your thoughts? Functionality too specialised to consider?

P.S. Parting thought: it is possible to "script" an entire performance so that the performer is simply cycling through a fixed sequence of presets. You could get through a whole show with just a 2 button controller, in theory. However, this isn't practical for at least 2 reasons:

1. It is very likely you need more than 128 total program changes in a performance.
2. Not all performances are predetermined; often, set lists change on-the-fly for various reasons.

**Comment by paradajz** [comment]
If I understood correctly, this is roughly what you want:

1. There is a new option for button message type called Program Change Offset
2. Putting the button in that mode and selecting the offset will just set internal offset for sending program change to some value without it sending any messages
3. If you have other buttons configured in program change mode, then the value should pass through configured offset as a final step and be sent incremented by that value. For example, if you set a button to set program change 10, and the offset is set to 4, the final sent value should be 14. Correct? What about the incrementing/decrementing type in that case? How do you expect the offset to influence those values?

**Reply by n3wkular** [reply]
Yes, I think we are in sync.

>For example, if you set a button to set program change 10, and the offset is set to 4, the final sent value should be 14. Correct?

Correct.

>What about the incrementing/decrementing type in that case? How do you expect the offset to influence those values?

You've just identified a UX issue with my suggested design, I think. A few thoughts:

- I believe there are two variables here: **1)** an "offset increment", which is global (or local to a logically grouped set of controls.) This variable can likely remain static after the user programs it.  **2)** The "offset value", which can be increased or decreased in increments of the "offset increment" at "runtime".
- I don't need to increment or decrement through the PCs 1-by-1 anymore, like I stated in the original post. To increment and decrement the _offset value_ in increments of the _offset increment_ is what is needed.
- For example, let's say the controller has 6 buttons: 2 buttons to increment and decrement the _offset value_, and 4 buttons to set PCs. So, the _offset increment_ value would be user-programmed to 4 to match the number of PC buttons.
- The 4 buttons to set PCs are programmed for PC#0 thru PC#3. So, when the _offset value_ is 0, the PCs aren't offset at all.
- However, the other 2 buttons allow the user to increase and decrease the _offset value_ in increments of the _offset increment_. So, increasing the _offset value_ once sets its value to 4, and the 4 PC buttons now set PC#4 thru PC#7 (the programmed PC# + _offset value_).
- Increasing again sets it to 8, and the 4 PC buttons set PC#8 thru PC#11, and so on.
- Decreasing it assigns the current _offset value_ as _offset value_ = (_offset value_ - _offset increment_)

I hope I am not over-complicating things, and I hope I am explaining clearly.

Cheers for reading this far! What do you think?

**Reply by paradajz** [reply]
Yeah I think it's quite clear actually. I'll need to do some refactoring first before I move on to this, but I think I could have a prototype maybe later today even.

**Reply by n3wkular** [reply]
Wow. That's totally unnecessary, but appreciated. It's not time-sensitive at all .I don't need this for a gig this weekend :grinning:

If you're interested in hacking at it and need anyone to test, I'd gladly help.

**Reply by paradajz** [reply]
Initial implementation is on master, feel free to try it out. You'll notice new message type for buttons in the UI: Program Change Offset. Select that and all further program change messages will be incremented by that offset (until it's changed).

**Reply by n3wkular** [reply]
Great work. I'll be able to try it out tomorrow morning EST. I haven't tried the Docker image for development yet either, so there will be a learning curve before I can build it.

Cheers and looking forward!

**Reply by n3wkular** [reply]
OK, a few things:
- The CI is failing its tests on the last build (but I assume you know that): https://github.com/shanteacontrols/OpenDeck/actions/runs/2464169720
- I've built the binaries and flashed them to a Mega 2560. However, in the Configurator, I do not see the new message type for buttons: Program Change Offset. (When you say "in the UI" above, you mean the Configurator?)
- The Configurator still shows v7.2.0 in the upper right hand corner. I assume the version is simply not bumped by a local build (and only bumped with a tag and a CI build or something like that?)

Just to be sure I built the right commit:
```
ubuntu@select-grunt:~/OpenDeck/src$ git status
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
ubuntu@select-grunt:~/OpenDeck/src$ git rev-parse --short HEAD
1008d3f8
ubuntu@select-grunt:~/OpenDeck/src$ git log -3
commit 1008d3f8e080c57d5388c599eb618ea4ae903541 (HEAD -> master, origin/master, origin/HEAD)
Author: paradajz <2544094+paradajz@users.noreply.github.com>
Date:   Wed Jun 8 21:53:43 2022 +0200

    app: add the ability to set offset to MIDIProgram

commit adbaddaf12ea9e6d46d942a3fcd7777ee0a691e9
Author: paradajz <2544094+paradajz@users.noreply.github.com>
Date:   Wed Jun 8 21:50:11 2022 +0200

    app: refactor io/common, use common util for inc/dec of values

commit 209fcc98b5eeb9963364a30d0d6d1c07497fc71e
Author: paradajz <2544094+paradajz@users.noreply.github.com>
Date:   Mon Jun 6 23:47:58 2022 +0200

    flash.sh: fix discovery of avr mcus
```
From there, the make script is successful with the mega2560 and mega16u2 targets. The resulting merged hex files are different (simple `cmp` check) from the v7.2 hex files posted on the releases page. But when flashed onto my board, there is no joy: everything works but the new button message type is not in the dropdown.

In the code, I can see `PROGRAM_CHANGE_OFFSET` added to the `messageType_t` enum, but I have no idea how the Configurator discovers this (and then translates it to a common English string in the UI.) 

Do you see anything obvious I've done wrong? How can I help troubleshoot?

Cheers!

**Reply by paradajz** [reply]
> The CI is failing its tests on the last build (but I assume you know that):

I went to sleep before it finished. It takes its merry time... I've broken regular program change and tests caught that. Fixed now, but not related to your problem.

> I've built the binaries and flashed them to a Mega 2560. However, in the Configurator, I do not see the new message type for buttons: Program Change Offset

It's the last option. If you don't see it try reloading the configurator without cache (Press F12, then right click on the reload button, then Empty Cache and Hard Reload). Now that I think of it, I should move this option below the existing Program change option group.

![image](https://user-images.githubusercontent.com/2544094/172779781-fbb89aea-33fe-4fc3-aef0-98941e640e8d.png)

> Just to be sure I built the right commit:

You did.

>  but I have no idea how the Configurator discovers this (and then translates it to a common English string in the UI.)

https://github.com/shanteacontrols/OpenDeckUI/commit/6583ff05fffb9f0105331fd1ebe2fe129dd144d3

**Reply by n3wkular** [reply]
> try reloading the configurator without cache

D'oh. I thought I had. :facepalm: 

> Now that I think of it, I should move this option below the existing Program change option group.

Yea, good idea.

Here, I am setting the message type to Program Change Offset for Button 10
![image](https://user-images.githubusercontent.com/107003685/172823634-3884c397-17ef-4644-9cad-9f65bc23ca63.png)
Log looks good to me.

Now I've moved to Button 11 and come back to Button 10
![image](https://user-images.githubusercontent.com/107003685/172824101-0e2e3a53-4b01-4ba7-af6b-6fa3fdf7825f.png)
Yep, the value is 4 (but it's querying that twice ... probably not important). Message type is 25 further down in the log. As expected.

At this point, pressing Button 10 on the controller does nothing. No MIDI message, no system activity registers, nothing. I think this is what you intended and this is just a prototype of the offset functionality.

Here I am programming Button 4 to send PC#4 and then pressing that button on the controller
![image](https://user-images.githubusercontent.com/107003685/172825150-3ed56ab6-b9e4-4c31-adc8-929f74394612.png)
Seems like Button 4 is sending a PC#**0** message, but we were expecting PC#**8**. Confirmed in a MIDI Monitor.
![image](https://user-images.githubusercontent.com/107003685/172825500-711f04bf-59b3-4d97-a052-398893dcbea6.png)
(Timestamps are different because I cleared the log and pressed Button 4 again.)

A few other observations:
- No matter what value Button 10 is set to, every PC is PC#0 globally.
- Did a factory reset. Programmed Button 4 for PC, and it still sends PC#0. Every PC is PC#0 globally, even without setting the PC Offset anywhere.

I realise this is just prototyping the offset maths, but is assigning the Program Change Offset value to a Button the best way to do this?
- In my mind, the offset value is a global variable, and we should have message types like Program Change Offset Inc/Dec that we can assign to controller Buttons.
- And these message types should have a value for how _much_ we want the Program Change Offset value to increment or decrement. (This is a little different than the idea of 2 variables [one for offset value, another for offset increment value] further up in this thread. Assigning the inc/dec value to Buttons with such inc/dec message types might be a better idea, IMO.)
- Of course, these message types are meaningless to any other device, so technically, there's no need to send out a message (the state change is only relevant to our controller and the internal Program Change Offset value.)

If all of that makes sense and we're on the same page, just ignore my blathering. Cheers again for taking the time to tackle this!

**Comment by paradajz** [comment]
> At this point, pressing Button 10 on the controller does nothing. No MIDI message, no system activity registers, nothing. I think this is what you intended and this is just a prototype of the offset functionality.

Yes. It only sets the offset.

> No matter what value Button 10 is set to, every PC is PC#0 globally.

This was actually a bug - are you on latest commit? It's fixed on master.

**Reply by n3wkular** [reply]
Seems to work now. In image below:
- Set B5 to type PCO and its value to 4
- Set B4 to type PC and its value to 4
- Pressed B5 on the controller
- Pressed B4 on the controller and PC#8 was sent
- Set B5's value to 10
- Pressed B4 and PC#14 was sent
![image](https://user-images.githubusercontent.com/107003685/172841322-65b5c3fc-ed5f-4392-9e3c-3f8f3f6539c6.png)

Awesome. So, now the offset maths work. Still thinking about the last bit of my message earlier:
>I realise this is just prototyping the offset maths, but is assigning the Program Change Offset value to a Button the best way to do this?
> - In my mind, the offset value is a global variable, and we should have message types like Program Change Offset Inc/Dec that we can assign to controller Buttons.
>  - And these message types should have a value for how much we want the Program Change Offset value to increment or decrement. (This is a little different than the idea of 2 variables [one for offset value, another for offset increment value] further up in this thread. Assigning the inc/dec value to Buttons with such inc/dec message types might be a better idea, IMO.)
> - Of course, these message types are meaningless to any other device, so technically, there's no need to send out a message (the state change is only relevant to our controller and the internal Program Change Offset value.)

Am I making any sense? Would a quick mockup help demonstrate what I am thinking?

**Comment by paradajz** [comment]
Ah, so the thing that's missing is the ability to press the offset for the 2nd (and 3rd, and 4th etc) time - right now if you configure a button in offset mode with offset being 4 for example, no matter how many times you press it it's not going to change further.

Right, so, to revise this, a following would be required:

* Message type Preset Change Offset Inc
* Message type Preset Change Offset Dec

And the flow:

* Set one button to Preset Change Offset Inc mode, offset is 4
* Press that button once -> offset 4 is set
* Press it again -> offset 8 is set
* Press it again -> offset 12 is set

Now, here comes the junction point:

* Configure another button in Dec mode, offset is 2
* Press that button
* What happens now with the current offset being 12? Is it set to 10, or, is it initially set to 2 and then on consecutive press to 0?

The same question actually goes if you have 2 inc buttons: set one offset to 2, and the other to 4. What happens on change? Immediate inc/dec or relative inc/dec depending on the current offset?

**Reply by n3wkular** [reply]
Yep. That's exactly what I am thinking ^^
![image](https://user-images.githubusercontent.com/107003685/172845845-40e88716-c86f-49bf-bc96-93fefbeda7cb.png)

Merging this mockup and your flow:
- When offset is 0, B1 sends PC#1 (because that's what it's programmed for in the Configurator)
- When offset is 4, B1 sends PC#5
- When offset 8, B1 sends PC#9

Re: the junction point: I hadn't really thought about the offset inc and offset dec being different values until I saw your first implementation. In this particular use case, I would set them to the same value (so I can cycle through all the AmpliTube presets in groups of 4.)

> What happens now with the current offset being 12? Is it set to 10, or, is it initially set to 2 and then on consecutive press to 0?

I'd say 10. Then 8 on next press, then 6, etc. Always relative, IMO. The feature is designed to navigate predictably through presets in software.

> The same question actually goes if you have 2 inc buttons: set one offset to 2, and the other to 4. What happens on change? Immediate inc/dec or relative inc/dec depending on the current offset?

That's more flexible than needed for my current use case, but my personal preference would be:
- Yes, you could set any button to be an offset inc or dec button.
- Any such button could be programmed to any value between 1 and 127.
- Anytime any such button is pressed, the change is always _relative_ to the current offset value.
- In any case, 0 should be the lowest possible offset value.

This would allow some fancier stuff than just my particular use case, but maybe that's over-engineering. Imagine an inc button set to 4 and a dec button set to 1. Maybe that's useful for someone, maybe not. If it isn't, then just set your inc/dec buttons with a common value. ¯\_(ツ)_/¯

Now it makes sense for visual feedback on the controller itself. Perhaps a screen that displays the current offset value (but let's not get ahead of ourselves :grin: )

Thoughts?

**Reply by n3wkular** [reply]
BTW, there's nothing magickal nor arbitrary about the number 4 I keep using in this use case. Many guitar modelling software "live" modes display the presets in groups of 4. Like this.
![image](https://user-images.githubusercontent.com/107003685/172851750-2d6af9dc-32be-480f-adf5-32b5c2ecfe8b.png)

**Reply by paradajz** [reply]
Pushed the changes both in UI and here:

* UI now shows Program Change Offset Increment and Program Change Offset Decrement
* The two new options are grouped now with the rest of the Program Change types
* Consecutive presses will now either decrement or increment the offset by specified amount

Let me know.

**Reply by n3wkular** [reply]
Just built and flashed the board. Based on some breadboard testing, it seems to be doing exactly what I was thinking. :metal: Cheers! I'll do a proper prototype soon.

Do you think this is a useful feature? And what's left to do from your side? If you decide to release this, can I help with the documentation?

And one last "cherry on top": it would be nice if there was some way to display the current Program Change Offset value on the controller. Perhaps in a small SSD1306 display. You know better how that would work; any high-level thoughts?

**Reply by paradajz** [reply]
> Just built and flashed the board. Based on some breadboard testing, it seems to be doing exactly what I was thinking. metal Cheers!

Great.

> Do you think this is a useful feature?

Doesn't really matter what I think. OpenDeck did everything I wanted years ago, but I love adding whatever other people find useful. 

> And what's left to do from your side? If you decide to release this, can I help with the documentation?

Probably some kind of indication of the offset, both on the display and through SysEx at least. Maybe in some other way as well? Not sure really.

**Reply by n3wkular** [reply]
> Doesn't really matter what I think.

Well, technically, as the maintainer, you decide what's noise or bloat and what is core to the vision for the project. :wink: But I'm glad you're so open to extensions and generous with your time.

With this new capability, I am refactoring the controller I wanted it for. Finding some LED visual feedback issues. E.g., it's easy to activate an LED above one of the buttons for a single PC#, but when the PC# of that button changes with the offset, things get tricky. I have a few ideas; will try and report back.

**Reply by paradajz** [reply]
> Well, technically, as the maintainer, you decide what's noise or bloat and what is core to the vision for the project. wink But I'm glad you're so open to extensions and generous with your time.

You can read my thoughts on this here: https://github.com/shanteacontrols/OpenDeck/discussions/153#discussioncomment-2666734
It just so happened that this feature seemed like a nice generic addition while being straightforward to implement.

> With this new capability, I am refactoring the controller I wanted it for. Finding some LED visual feedback issues. E.g., it's easy to activate an LED above one of the buttons for a single PC#, but when the PC# of that button changes with the offset, things get tricky. I have a few ideas; will try and report back.

Please do.

**Reply by n3wkular** [reply]
OK, this will be all for me today. Here's what I'd like to do but can't find a way to do it.
![image](https://user-images.githubusercontent.com/107003685/172911286-2fc8a862-210b-4cb8-a6a2-ca557c92f1e4.png)

- A way to visually indicate the current offset value on the controller would be great. I've never programmed MIDI or SysEx directly, but I can envision a way to do this with a second microcontroller if OpenDeck would respond to a SysEx request for the current offset value. You probably have a much better idea, however.
- No matter what the offset value is, I'd like a way to activate each LED in the mockup when the button beneath it is pressed (indicating which of the 4 presets the software was instructed to load). At the moment, I don't think that's possible because an LED can only be mapped to a single Activation ID (if the LED is set to Control Type > Program Change). But the PC# values change for these buttons when the offset is non-zero. :thinking: You know MIDI in general far better than I do. Any ideas?

- [x] Officially enter "Wouldn't it be nice?" territory.

Cheers once again.

**Reply by gh0stless** [reply]
As Guitarrrig runs on a PC you maybe can use a small helpertool as filter together with an virtual midi cable. The helpertool can programmed in Max/MSP or with C++ and Juce Framework for example.

**Reply by paradajz** [reply]
> A way to visually indicate the current offset value on the controller would be great.

This can be added to OLED display. Unfortunately my testing OLED display broke and I'm still waiting for the new one so it's going to be a while before I can do anything display related.

> a way to do this with a second microcontroller if OpenDeck would respond to a SysEx request for the current offset value. You probably have a much better idea, however.

No need for second MCU at all. It can be handled by OpenDeck just fine.

> No matter what the offset value is, I'd like a way to activate each LED in the mockup when the button beneath it is pressed (indicating which of the 4 presets the software was instructed to load). At the moment, I don't think that's possible because an LED can only be mapped to a single Activation ID (if the LED is set to Control Type > Program Change). But the PC# values change for these buttons when the offset is non-zero. thinking You know MIDI in general far better than I do. Any ideas?

LEDs can be configured to indicate the specific program change. Not sure what the problem is? When you change the offset the LEDs would be out of sync. This should be solved, yeah. Need to add some tests for offset functionality in the next few days anyways.

**Reply by n3wkular** [reply]
>No need for second MCU at all. It can be handled by OpenDeck just fine.

OK. I need to mess around with a display because I hardly know what I am talking about. In the Configurator, I didn't see a way to display the Program Change Offset value. Perhaps you're suggesting you'd have to add the option?

>LEDs can be configured to indicate the specific program change. Not sure what the problem is?

I could have been more clear, or I am missing something obvious. But the word "specific" ^^ is a clue for me. Let's use the mockup again and pretend we're starting with a Program Change Offset value of 0.
![image](https://user-images.githubusercontent.com/107003685/172948430-68923819-31a0-4456-9c11-6404a4379d5a.png)

The 1 button is configured like this:
![image](https://user-images.githubusercontent.com/107003685/172949839-9f76c42d-daff-4589-8c6e-c2396c6d0ee1.png)

And the LED directly above the 1 button is configured like this:
![image](https://user-images.githubusercontent.com/107003685/172950023-b6adba5d-1e51-4a8d-8539-e686d2782a5e.png)

The LED is supposed to indicate which of the software presets the 4 buttons _currently_ represent is loaded. And that works fine when the Program Change Offset is 0. Pressing the 1 button results in a PC#1, the preset loads, the LED activates.

Now:
- Pressing the UP button increments the Program Change Offset value to 4.
- At this point, the LED above button 1 is still activated; not a huge deal because the user just pressed the UP button in order to select a new preset, but I see your point about synchronisation.
- The user decides they want preset 5, so they press the 1 button again (they know 4 + 1 = 5).
- The LED above button 1 _deactivates_ because the message sent was PC#5 but the LED's Activation ID is _specifically_ 1.
- Preset 5 loads in the software.

At that point I don't understand how one would dynamically assign an Activation ID of 5 to the LED above the 1 button. In my mind, there's no way to activate that LED unless the Program Change Offset value is 0 and the user presses the 1 button.

Hope that helps. Cheers.

P.S. It occurred to me while hanging new gutters on my house a few hours ago: what was just implemented is essentially how a keyboard controller shifts the octave up and down. The keys all do the same thing, just offset by an octave. I don't know. Maybe that's stupid. :smiley:

**Reply by n3wkular** [reply]
@gh0stless 
> As Guitarrrig runs on a PC you maybe can use a small helpertool as filter together with an virtual midi cable. The helpertool can programmed in Max/MSP or with C++ and Juce Framework for example.

Good idea and thanks for the suggestion. I'm holding out for a more elegant solution as I might be running AmpliTube on an iPad, or my Android phone from time to time. Cheers.

**Reply by paradajz** [reply]
> Perhaps you're suggesting you'd have to add the option?

Yes.

> The LED is supposed to indicate which of the software presets the 4 buttons _currently_ represent is loaded. And that works fine when the Program Change Offset is 0. Pressing the 1 button results in a PC#1, the preset loads, the LED activates.
> 
> Now:
> 
> * Pressing the UP button increments the Program Change Offset value to 4.
> * At this point, the LED above button 1 is still activated; not a huge deal because the user just pressed the UP button in order to select a new preset, but I see your point about synchronisation.
> * The user decides they want preset 5, so they press the 1 button again (they know 4 + 1 = 5).
> * The LED above button 1 _deactivates_ because the message sent was PC#5 but the LED's Activation ID is _specifically_ 1.
> * Preset 5 loads in the software.
> 

In my mind, the current behavior is actually correct. Keep in mind that in the firmware, buttons, LEDs, encoders etc. are separate concepts, not really tightly related to one another. The offset works on buttons module, not for LEDs (currently). If you configure the LED to display Program 1, then in my mind it should be on only on Program 1. Offset is a currently just a button concept. You could theoretically have 4 buttons for programs and 2 more for offset, but then have 20 LEDs for Program Change, each indicating one. Having them react in accordance with the offset, which is a button thing, could be confusing. There's no obviously correct behavior here, but I'm leaning heavily towards the current one. Maybe I could add an *option* for LEDs to listen to offset event  and then adjust the indication accordingly, but I certainly wouldn't leave it on by default. What do you think?

**Reply by n3wkular** [reply]
>Offset is a currently just a button concept. You could theoretically have 4 buttons for programs and 2 more for offset, but then have 20 LEDs for Program Change, each indicating one. Having them react in accordance with the offset, which is a button thing, could be confusing.

100% agree when looking at it from the perspective of the firmware. But if we think about the UX of the type of _controller_ I am trying to build, it is confusing if:
- The controller's primary function enables the user to navigate through 128 Program Changes in software that groups them by 4
- There are 4 LEDs, each logically corresponding to one of 4 PC buttons
- The LEDs only ever activate for PC#1 through PC#4
- There's no visual feedback for 97% of the Program Changes

I could add 128 LEDs to the controller, but that's like hunting squirrels with hand grenades, IMO. And I don't think it would be any less confusing, because the magick number 4 nicely corresponds with how the software groups and displays its presets. (See AmpliTube screenshot above ^^). I could also omit the LEDs completely, but I don't like that UX either: no visual feedback at all.

>Maybe I could add an option for LEDs to listen to offset event and then adjust the indication accordingly

This would enable the use case I am trying to implement. That would be awesome.

>but I certainly wouldn't leave it on by default. What do you think?

I totally agree. IMO, the offset, new button message types, and LEDs that integrate with those concepts are all fairly specialised. None of these features should be the default behaviour, but they all work really well together for controllers that need it. Check out this commercial product: https://www.ikmultimedia.com/products/irigstompio/images/1.0/iRigStompIO_iPad_dx34@2x.jpg. It does exactly what I am trying to make my controller do (except it requires the user to hold a button to change the offset; I like what you just implemented better.) The LEDs honour the offset so they mirror what is loaded in the software GUI.

In my mind, while OpenDeck is a super powerful platform for people building MIDI controllers, the real end users are the _people using the controllers_. _Those_ end users might not even know what the heck OpenDeck _is_; controller UX is their main motivation for judging a controller useful or not. And FWIW, there are lots of forum posts where people are trying to find controllers just like this for various guitar rig software, and [the offset is the thing they can't do](https://cgi.ikmultimedia.com/ikforum/viewtopic.php?f=5&t=27169&p=130879#p130879). If OpenDeck can do it with specialised, elegant UX, then it fills a market need. I've already started telling guitar players about it :wink:

Cheers again for an awesome platform!

**Reply by paradajz** [reply]
> This would enable the use case I am trying to implement. That would be awesome.

Added the new option for LEDs. It's visible now in the UI. Just reflash the board. Note that the latest changes will wipe your config since I've added new database parameter. Don't have a setup currently to try this out, so let me know.

> I've already started telling guitar players about it wink

Much appreciated!

**Reply by n3wkular** [reply]
That is GREAT! However, it will be a few days before I can test. Headed out of town for the weekend. Enjoy yours and cheers!

**Reply by paradajz** [reply]
Added the tests, everything is passing, also tested it out on my physical test setup, seems to work fine. Will release the new version after confirmations from your side. Possibly help text in the UI needs some tweaking... let me know on that next week.

> If you decide to release this, can I help with the documentation?

Would you be interested in writing a guest blog post on my site on what kind of controller you're building and how these new features helped you in making it possible?

**Reply by n3wkular** [reply]
OK, everything is working as I'd expect. I placed an order for parts needed to complete the physical prototype; they should arrive later this week.

I don't think the help text needs much work at all, honestly. But that might be because I understand the use case already. What's more important, IMO, is documenting how the new options work together; it's niche and a little complicated. So, a blog post, wiki page, or both? Maybe I can write the blog post, we'll review it, and then take the salient points and publish a wiki page. What do you think?

Cheers again!

**Reply by paradajz** [reply]
Glad it works out for you.

The documentation for buttons:
https://github.com/shanteacontrols/OpenDeck/wiki/Configurable-features#type

And for LEDs:
https://github.com/shanteacontrols/OpenDeck/wiki/Configurable-features#global-parameters

As for the blog post, you can open PR on [website repo](https://github.com/shanteacontrols/website) with the only change being newly added Markdown-formatted blog post (website is hosted via GitHub and gets rebuilt and deployed on each push). You can also add some images. See for example how I've written blog post announcing the new boards:

https://github.com/shanteacontrols/website/commit/8eef0a69b9da20acdf6812cea9a239cf9f89a417

**Comment by n3wkular** [comment]
Brilliant. Yes and yes. Will get back to you tomorrow EST.

**Comment by paradajz** [comment]
The release is out: https://github.com/shanteacontrols/OpenDeck/releases/tag/v7.3.0

**Reply by n3wkular** [reply]
:+1:  All the doc looks good; great work. No need to tweak anything unless people are confused about how to use it. I'll submit a PR with a blog post sometime this week. I'd prefer to have pictures of a physical prototype available.

**Reply by n3wkular** [reply]
Real life strikes: had a family emergency the last few days. Haven't forgotten and will be able to write something up by early next week.

**Reply by paradajz** [reply]
Hey no worries. Hope everything's fine.
